### DISCUSSION TITLE
How to close a modal from a different component?

### DISCUSSION THREAD

**Discussion by multiplehats** [vnphanquang/svelte-put#113] [Q&A] [answered]
Have set up my modal as described in the documentation. Found myself in a situation where I need to close that particular modal from a component that lives way down the tree and has no access to it whatsoever.

I'm currently doing this (and it works), but I don't think this is the right way to do it. Because `pushed` is undefined. What am I missing.

```svelte
<script lang="ts">
	import type WebsiteModal from '$components/website/WebsiteModal.svelte';

	let pushed: ModalPushOutput<WebsiteModal>;

	function closeModal() {
		console.log(pushed);
		modalStore.pop(pushed);
	}
</script>

<button
	type="button"
	on:click={closeModal}
	class="inline-flex items-center p-2 border border-slate-300 bg-red-400  shadow-sm text-sm font-medium rounded-lg text-slate-900 bg-white hover:bg-slate-100 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary-500"
>
	<span class="sr-only">Return</span>
	<X class="h-5 w-5" aria-hidden="true" />
</button>
```

**Comment by vnphanquang** [comment, marked as answer]
@multiplehats hello there, good question. When calling `pop(undefined)`, the topmost modal on the stack will close. If you know that the modal you're trying to `pop` is always on top, then all is good.

To close a specific modal in the middle of the stack, you need to pass id of the pushed modal:

```typescript
modalStore.pop({ id: 'some-thing' } as ModalPushOutput<WebsiteModal>);
```

This id is returned in the `ModalPushOutput`: 

```typescript
const { id } = modalStore.push(SomeComponent) // id here is generated for you;
// or
const { id } = modalStore.push({ id: 'user-defined-id', component: SomeComponent }); // id here is same as specified
```

So you basically have to keep this id somewhere after `push` (svelte context, another store, ...) and use it in `pop` where needed.

**Reply by multiplehats** [reply]
Ah, I see. Beautiful, that's all I needed to know. 

Thanks for your prompt reply!

**Reply by vnphanquang** [reply]
I should make the interface a bit more user-friendly here so that the ugly `as` phrase can be omitted eh? Problem is this is to infer the `pop` return value.

```typescript
const popped = modalStore.pop(pushed) // type of popped here is inferred;
```

**Reply by multiplehats** [reply]
I really like the whole API. I'm kinda using it as a headless UI. 

if I'm honest, I would rather see it being headless UI first, rather than provide a basic modal out of the box. 

But I'm using af few things here and there from this repo, and your work is amazing!

**Reply by multiplehats** [reply]
If you have a custom modal, how does it call resolve after you use the following @vnphanquang :

```ts
modalStore.pop({ id: 'some-thing' } as ModalPushOutput<WebsiteModal>);
```

Have omitted a lot of my code for brevity but here it is:

```svelte
<script lang="ts">
	import clsx from 'clsx';
	import { fly } from 'svelte/transition';
	import { clickoutside } from '@svelte-put/clickoutside';
	import { createEventDispatcher, onMount, SvelteComponent } from 'svelte';
	import type { ResolveTrigger } from '@svelte-put/modal';
	import type { Website } from '$sanity/types';
	import { shortcut } from '@svelte-put/shortcut';
	import { browser } from '$app/environment';
	import type WebsiteModalContent from './WebsiteModalContent.svelte';
	import { page } from '$app/stores';
	import { link } from '$lib/config/links';

	const dispatch = createEventDispatcher<{
		resolve: {
			trigger: ResolveTrigger;
			payload: string;
		};
	}>();

	export let id: Website['id'];
	export let slug: Website['slug'];
	export let escape = true;
	export let disableScroll = true;

	if (!id) {
		throw new Error('WebsiteModal: id is required');
	}

	let lazyElement: typeof WebsiteModalContent | null = null;
	let website: Website | null = null;
	let parent: HTMLElement;

	function resolve(trigger: ResolveTrigger, payload: string) {
		document.body.classList.remove('overflow-hidden');
		dispatch('resolve', { trigger: 'clickoutside', payload: 'completely custom modal' });
	}

	function onPressEscape() {
		resolve('escape', '');
	}

	function onClickOutside() {
		resolve('clickoutside', '');
	}

	$: if (disableScroll && browser) {
		document.body.classList.add('overflow-hidden');
	}
</script>

<div
	bind:this={parent}
	class={clsx(
		'z-50 p-0 items-end fixed inset-0 flex justify-center min-h-full h-auto bg-black/70 overflow-x-hidden overflow-y-auto'
	)}
>
	<!-- Modal wrapper -->
	<div
		use:clickoutside={{ limit: { parent } }}
		on:clickoutside={onClickOutside}
		in:fly={{ y: 300, duration: 400 }}
		class={clsx('w-full h-full relative')}
	>
		<!-- Modal content -->
		{#if lazyElement && website}
			<svelte:component this={lazyElement} {website} />
		{/if}
	</div>
</div>
```

My problem is dat resolve is not being called, thus `document.body.classList.remove('overflow-hidden')` isn't called

**Reply by vnphanquang** [reply]
Hmm currently, a modal when popped manually will go through a `resolve` function internal to `store` and not the one defined in the component. In other words you have to do the side effects manually where `pop` happens.

This is a problem. Sorry for the lack of support here. I will see what I can do.

And for the record I agree too `@svelte-put/modal` should have been headless first.

**Reply by multiplehats** [reply]
Noted! And yeah, no need to apologize — I don't expect anything this is OSS after all.

Appreciate your time!

**Reply by multiplehats** [reply]
Hmm. Not sure what a clean way would be to handle the resolve side effect.

Do you have anything top of mind @vnphanquang? When I subscribe to the appStore in my modal, there isn't really anything I can use that indicates it has been closed

**Reply by vnphanquang** [reply]
@multiplehats looking closely I don't think it is possible to support side effects as in the code above. `resolve` there is a locally-defined function that is not inspect-able at runtime (without some hacking at least).  

For this though

> Do you have anything top of mind @vnphanquang? When I subscribe to the appStore in my modal, there isn't really anything I can use that indicates it has been closed

I've also been thinking. Something like one of these maybe

```typescript
modalStore.push({
  component: SomeComponent,
  onPop: () => {},
});
```

```typescript
modalStore.on('pop', modalPushedId, () => {});
// or
modalStore.onPop(modalPushedId, () => {});
```

From a user perspective, do you have any suggestion?

**Reply by vnphanquang** [reply]
Those above have to be handled separately from the component code though, which I don't quite like.

For what it's worth, when a modal is popped, it is `unmounted`, so svelte `onMount` can be used here

```typescript
import { onMount } from 'svelte';

onMount(() => {
  return () => {
    // code run when modal is popped
  };
});
```

**Reply by multiplehats** [reply]
These are good suggestions.

I guess I don't like
```
modalStore.push({
  component: SomeComponent,
  onPop: () => {},
});
```

As much, because I personally would rarely declare the onPop function where I would also the declare the `modalStore.push`.

So for me, `modalStore.onPop(modalPushedId, () => {});` this would feel really natural. And I can have different logic throughout the app that handles the closing differently.

Turns out I really need this lol 😭

**Reply by vnphanquang** [reply]
`modalStore.onPop(modalPushedId, () => {})` sounds fine to me. I understand, side effect is always the culprit. I'll add this in a patch release shortly. Hang tight

**Reply by multiplehats** [reply]
Agreed! 

Beautiful, thank you very much.

**Reply by vnphanquang** [reply]
Pending patch release [0.6.1](https://github.com/vnphanquang/svelte-put/blob/main/packages/misc/modal/CHANGELOG.md#061) has added this support. You can find [example & docs here](https://github.com/vnphanquang/svelte-put/blob/main/packages/misc/modal/api/docs/modal.modalstore.onpop.md) (it's basically what we discussed).

Unfortunately seems that npm is down, packages can't be published at the moment: https://status.npmjs.org/. Will update you once it's up @multiplehats

**Reply by multiplehats** [reply]
> Pending patch release [0.6.1](https://github.com/vnphanquang/svelte-put/blob/main/packages/misc/modal/CHANGELOG.md#061) has added this support. You can find [example & docs here](https://github.com/vnphanquang/svelte-put/blob/main/packages/misc/modal/api/docs/modal.modalstore.onpop.md) (it's basically what we discussed).
> 
> 
> 
> Unfortunately seems that npm is down, packages can't be published at the moment: https://status.npmjs.org/. Will update you once it's up @multiplehats 
> 
> 

Dude this is amazing. Thanks for picking this up so quickly 🚀👏

**Reply by vnphanquang** [reply]
All good now that npm team has resolved the incident. Let me know if have any further issue.

[Ref latest release](https://github.com/vnphanquang/svelte-put/releases/tag/%40svelte-put%2Fmodal%400.6.1)
